op-supervisor/supervisor/backend/syncnode/reset.go

```diff
@@ -0,0 +1,187 @@
+
+	var z eth.BlockID
+
+	// check if the last unsafe block we saw is consistent with the logs db
+	err := m.backend.IsLocalUnsafe(ctx, m.chainID, m.lastNodeLocalUnsafe)
+	if errors.Is(err, types.ErrConflict) {
+		m.log.Warn("local unsafe block is inconsistent with logs db. Initiating reset",
+			"lastUnsafeblock", m.lastNodeLocalUnsafe,
+			"err", err)
+		z = m.lastNodeLocalUnsafe
+	}
+
+	// check if the last safe block we saw is consistent with the local safe db
+	err = m.backend.IsLocalSafe(ctx, m.chainID, m.lastNodeLocalSafe)
+	if errors.Is(err, types.ErrConflict) {
+		m.log.Warn("local safe block is inconsistent with logs db. Initiating reset",
+			"lastSafeblock", m.lastNodeLocalSafe,
+			"err", err)
+		z = m.lastNodeLocalSafe
+	}
+
+	// there is inconsistency. initiate reset
+	if z != (eth.BlockID{}) {
+		m.resetting.Store(true)
+		m.cancelReset.Store(false)
+		m.ongoingReset = &ongoingReset{a: eth.BlockID{}, z: z}
+		// trigger the reset asynchronously
+		go m.prepareReset()
+	} else {
+		m.log.Debug("no inconsistency found")
+		m.resetting.Store(false)
+		m.ongoingReset = nil
+	}
+}
+
+// prepareReset prepares the reset by bisecting the the search range
+// until the last consistent block is found. It then identifies the correct
+// unsafe, safe, and finalized blocks to target for the reset.
+func (m *ManagedNode) prepareReset() {
+	internalCtx, iCancel := context.WithTimeout(m.ctx, internalTimeout)
+	defer iCancel()
+	nodeCtx, nCancel := context.WithTimeout(m.ctx, nodeTimeout)
+	defer nCancel()
+	// repeatedly bisect the range until the last consistent block is found
+	for {
+		if m.cancelReset.Load() {
+			m.log.Info("reset cancelled")
+			m.resetting.Store(false)
+			m.ongoingReset = nil
+			return
+		}
+		if m.ongoingReset.a.Number >= m.ongoingReset.z.Number {
+			m.log.Error("no reset target found. restarting reset",
+				"a", m.ongoingReset.a,
+				"z", m.ongoingReset.z)
+			// check consistency again once this function returns
+			// the reset failed, and may need to be retried
+			defer m.checkConsistencyWithDB()
+			return
+		}
+		if m.ongoingReset.a.Number+1 == m.ongoingReset.z.Number {
+			break
+		}
+		err := m.bisectRecoveryRange(internalCtx, nodeCtx)
+		if err != nil {
+			m.log.Error("failed to bisect recovery range", "err", err)
+			defer m.checkConsistencyWithDB()
+			return
+		}
+	}
+
+	// the bisection is now complete. a is the last consistent block, and z is the first inconsistent block
+	target := m.ongoingReset.a
+	m.log.Info("reset point has been found", "target", target)
+	var unsafe, safe, finalized eth.BlockID
+
+	// the unsafe block is always the last block we found to be consistent
+	unsafe = target
+
+	// the safe block is either the last consistent block, or the last safe block, whichever is earlier
+	lastSafe, err := m.backend.LocalSafe(internalCtx, m.chainID)
+	if err != nil {
+		m.log.Error("failed to get last safe block. cancelling reset", "err", err)
+		defer m.checkConsistencyWithDB()
+		return
+	}
```

Comment:
Does reset overwrite cross-safe or local-safe? I thought it was changing cross-safe. If cross-safe, then this shouldn't be using local-safe here.

I think it makes sense to continue the derivation from the local-safe point, but we should be careful to not overwrite cross-safe then. And there's still a cross-safe thing that can be skipped forwards to. So maybe we should be using both in the reset call.

Shall we change the reset-API typing to be more explicit about the local-safe/cross-safe difference?

Reply:
Good question, I had assumed Local Safe, but we should check and make it explicit, yes.